### ISSUE TITLE
pkg.latest and pkg.installed do not respect apt pinning

### ISSUE THREAD

**Issue by boltronics** [saltstack/salt#36363]
The latest version of Salt (which I have packaged as version `2016.3.3-1`) has a critical problem (https://github.com/saltstack/salt/issues/36321) which I only noticed in our staging environment. I need to roll it back.

I can't just use:

```
salt-minion:
  pkg.installed:
    - version: 2016.3.1-1
```

because salt-minion 2016.3.3-1 has a dependency on salt-common = 2016.3.3-1, and apt-get won't downgrade a package by default. We could instead run:

```
salt-minion:
  pkg.installed:
    - pkgs:
      - salt-minion: 2016.3.1-1
      - salt-common: 2016.3.1-1
```

That is undesirable since salt-minion can later be removed and we will not get the apt-get warning `The following package was automatically installed and is no longer required: salt-common` afterwards. A minor detail. A more important problem is that the next `pkg.upgrade` or `apt-get upgrade` call will re-upgrade `salt-minion` to the broken version again.

So to solve this, we need to pin the packages. Hence I created another state which creates a file with the following contents in `/etc/apt/preferences.d/salt-release`:

```
Package: salt-cloud salt-common salt-master salt-minion salt-syndic
Pin: version 2016.3.1-1
Pin-Priority: 1000
Explanation: Don't upgrade Salt packages.
```

Now when I run `apt-get upgrade`, I get:

```
The following packages will be upgraded:
  libmysqlclient18 mysql-common
The following packages will be DOWNGRADED:
  salt-common salt-minion
2 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 0 not upgraded.
Need to get 3,612 kB/4,366 kB of archives.
After this operation, 163 kB disk space will be freed.
Do you want to continue? [Y/n]
Reading changelogs...
apt-listchanges: Mailing serveradmin@sitepoint.com: apt-listchanges: changelogs for ip-172-31-21-108
(Reading database ... 39768 files and directories currently installed.)
Preparing to unpack .../salt-minion_2016.3.3-1_all.deb ...
[ ok ] Stopping salt minion control daemon: salt-minion.
Unpacking salt-minion (2016.3.3-1) over (2016.3.1-1) ...
...etc.
```

The pinning is working perfectly. However, we don't want to ssh into each machine and run `apt-get upgrade`. Maybe we can try `pkg.upgrade`?

```
$ sudo salt-call pkg.upgrade
[WARNING ] /usr/lib/python2.7/dist-packages/salt/grains/core.py:1493: DeprecationWarning: The "osmajorrelease" will be a type of an integer.

[INFO    ] Executing command ['apt-get', '-q', 'update'] in directory '/root'
[INFO    ] Executing command ['dpkg-query', '--showformat', '${Status} ${Package} ${Version} ${Architecture}\n', '-W'] in directory '/root'
[INFO    ] Executing command ['apt-get', '-q', '-y', '-o', 'DPkg::Options::=--force-confold', '-o', 'DPkg::Options::=--force-confdef', 'upgrade'] in directory '/root'
[ERROR   ] Command '['apt-get', '-q', '-y', '-o', 'DPkg::Options::=--force-confold', '-o', 'DPkg::Options::=--force-confdef', 'upgrade']' failed with return code: 100
[ERROR   ] stdout: Reading package lists...
Building dependency tree...
Reading state information...
The following package was automatically installed and is no longer required:
  python-systemd
Use 'apt-get autoremove' to remove it.
The following packages will be upgraded:
  libmysqlclient18 mysql-common
The following packages will be DOWNGRADED:
  salt-common salt-minion
2 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 0 not upgraded.
Need to get 3612 kB/4366 kB of archives.
After this operation, 163 kB disk space will be freed.
E: There are problems and -y was used without --force-yes
[ERROR   ] retcode: 100
[INFO    ] Executing command ['dpkg-query', '--showformat', '${Status} ${Package} ${Version} ${Architecture}\n', '-W'] in directory '/root'
local:
    ----------
    changes:
        ----------
    comment:
        Reading package lists...
        Building dependency tree...
        Reading state information...
        The following package was automatically installed and is no longer required:
          python-systemd
        Use 'apt-get autoremove' to remove it.
        The following packages will be upgraded:
          libmysqlclient18 mysql-common
        The following packages will be DOWNGRADED:
          salt-common salt-minion
        2 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 0 not upgraded.
        Need to get 3612 kB/4366 kB of archives.
        After this operation, 163 kB disk space will be freed.
        E: There are problems and -y was used without --force-yes
    result:
        False
$
```

Hmm... okay. Well `sys.doc pkg.upgrade` doesn't say anything about having a `force_yes` option, but it does mention a `dist_upgrade` option. Let's try that:

```
$ sudo salt-call pkg.upgrade dist_upgrade=True
[WARNING ] /usr/lib/python2.7/dist-packages/salt/grains/core.py:1493: DeprecationWarning: The "osmajorrelease" will be a type of an integer.

[INFO    ] Executing command ['apt-get', '-q', 'update'] in directory '/root'
[INFO    ] Executing command ['dpkg-query', '--showformat', '${Status} ${Package} ${Version} ${Architecture}\n', '-W'] in directory '/root'
[INFO    ] Executing command ['apt-get', '-q', '-y', '-o', 'DPkg::Options::=--force-confold', '-o', 'DPkg::Options::=--force-confdef', 'dist-upgrade'] in directory '/root'
[ERROR   ] Command '['apt-get', '-q', '-y', '-o', 'DPkg::Options::=--force-confold', '-o', 'DPkg::Options::=--force-confdef', 'dist-upgrade']' failed with return code: 100
[ERROR   ] stdout: Reading package lists...
Building dependency tree...
Reading state information...
The following package was automatically installed and is no longer required:
  python-systemd
Use 'apt-get autoremove' to remove it.
The following packages will be DOWNGRADED:
  salt-common salt-minion
0 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 0 not upgraded.
Need to get 0 B/3612 kB of archives.
After this operation, 160 kB disk space will be freed.
E: There are problems and -y was used without --force-yes
[ERROR   ] retcode: 100
[INFO    ] Executing command ['dpkg-query', '--showformat', '${Status} ${Package} ${Version} ${Architecture}\n', '-W'] in directory '/root'
local:
    ----------
    changes:
        ----------
    comment:
        Reading package lists...
        Building dependency tree...
        Reading state information...
        The following package was automatically installed and is no longer required:
          python-systemd
        Use 'apt-get autoremove' to remove it.
        The following packages will be DOWNGRADED:
          salt-common salt-minion
        0 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 0 not upgraded.
        Need to get 0 B/3612 kB of archives.
        After this operation, 160 kB disk space will be freed.
        E: There are problems and -y was used without --force-yes
    result:
        False
$
```

It's not happening. That's got to be a bug, but no matter since ideally we want this defined and actioned via a state anyway.

```
$ sudo salt-call -l debug state.sls salt.minion
...
[DEBUG   ] Rendered data from file: /var/cache/salt/minion/files/base/salt/minion.sls:
...
salt-minion:
  pkg:
    - latest
...
[INFO    ] Running state [salt-minion] at time 18:24:59.185045
[INFO    ] Executing state pkg.latest for salt-minion
[INFO    ] Executing command ['apt-cache', '-q', 'policy', 'salt-minion'] in directory '/root'
[INFO    ] Package salt-minion is already up-to-date
[INFO    ] Completed state [salt-minion] at time 18:24:59.197282 duration_in_ms=12.237
...
----------
          ID: salt-minion
    Function: service.running
      Result: True
     Comment: The service salt-minion is already running
     Started: 18:24:59.215809
    Duration: 71.76 ms
     Changes:   
...
```

Okay, so it didn't do anything after checking the package policy. Let's see what that's doing:

```
$ sudo apt-cache -q policy salt-minion
salt-minion:
  Installed: 2016.3.3-1
  Candidate: 2016.3.1-1
  Package pin: 2016.3.1-1
  Version table:
 *** 2016.3.3-1 1000
        500 http://my_repo_host/debian/ jessie/main amd64 Packages
        100 /var/lib/dpkg/status
     2016.3.1-1 1000
        500 http://my_repo_host/debian/ jessie/main amd64 Packages
     2016.3.0+ds-1~bpo8+1 1000
        100 http://ftp.us.debian.org/debian/ jessie-backports/main amd64 Packages
     2014.1.13+ds-3-abolte 1000
        500 http://my_repo_host/debian/ jessie/main amd64 Packages
        500 http://my_repo_host/debian/ jessie/main amd64 Packages
     2014.1.13+ds-3 1000
        500 http://http.debian.net/debian/ jessie/main amd64 Packages
$
```

It clearly says the candidate and pinned version is 2016.3.1, and yet it did nothing! Replacing `pkg.latest` with `pkg.installed` doesn't help either.

Other info:

```
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.2
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.3
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.0
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.1
            ZMQ: 4.0.5

System Versions:
           dist: debian 8.5 
        machine: x86_64
        release: 3.16.0-4-amd64
         system: Linux
        version: debian 8.5 
```

Basically, to roll back all Salt versions from the master, I'll need to run:
`$ sudo salt '*' cmd.run 'apt-get -q -y --force-yes -o DPkg::Options::=--force-confold -o DPkg::Options::=--force-confdef upgrade'`

This can't be right!

**Comment by gtmanfred** [comment]
Thank you for reporting this. I have been able to reproduce this problem.

**Comment by disaster123** [comment]
+1

**Comment by rehmanzile** [comment]
+1

**Comment by rehmanzile** [comment]
Any ETA on this issue?

**Comment by gtmanfred** [comment]
There is not, if you wanted to submit a PR we would greatly appreciate it.

But we currently have a lot of stuff ahead of this on the todo list.

**Comment by boltronics** [comment]
I am not aware of any change in behaviour, however downgrading packages via apt pinning isn't a common occurrence. I don't have time to test against the latest Salt right at the moment.

**Comment by twangboy** [comment]
Closing this issue due to age and lack of activity. Please test this on version 3006.2 and create a new issue if the problem persists. The new issue template has more information and will allow us to track and reproduce the issue more effectively. Thanks!
